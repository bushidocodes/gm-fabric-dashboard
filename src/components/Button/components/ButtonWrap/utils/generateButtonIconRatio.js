/**
 * Helper function that maps button sizes to numbers representing icon ratios
 *
 * @param {string} size
 * @returns number
 */
function mapSizeToButtonIconRatio(size) {
  switch (size) {
    case "xs":
      return 0.5;
    case "sm":
      return 0.75;
    case "lg":
      return 1.25;
    case "xl":
      return 1.5;
    case "normal":
    default:
      return 1;
  }
}

/**
 * Utility function that generates a CSS mixin containing button icon rulesets
 *
 * @param {string} size
 * @returns string
 */
function generateButtonIconRatio(size) {
  const viewboxSize = 24; // default viewbox size 24px
  const ratio = mapSizeToButtonIconRatio(size); // default ratio 1
  const viewboxDimensions = viewboxSize * ratio;
  return `> svg { width: ${/* sc-value */ viewboxDimensions}px; 
  height: ${/* sc-value */ viewboxDimensions}px; }`;
}

export default generateButtonIconRatio;
