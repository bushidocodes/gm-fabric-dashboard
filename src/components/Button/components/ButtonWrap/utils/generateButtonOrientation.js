import { spacingScale } from "style/styleFunctions";

/**
 * Utility function that generates a CSS mixin containing button orientation rulesets
 *
 * @param {string} [alignment="horizontal"]
 * @param {number} [spacingBase=0.5]
 * @returns string
 */
function generateButtonOrientation(
  alignment = "horizontal",
  spacingBase = 0.5
) {
  if (alignment === "horizontal") {
    return `
      flex-direction: row;
      > :not(:first-child) {
        margin-left: ${/* sc-value */ spacingScale(spacingBase)};
        flex-basis: auto;
      }
    `;
  } else if (alignment === "vertical") {
    return `
      flex-direction: column;
      padding-top: ${/* sc-value */ spacingScale(spacingBase)};
      > :not(:first-child) {
        margin-top: ${/* sc-value */ spacingBase / 2}px;
      }
    `;
  }
}

export default generateButtonOrientation;
