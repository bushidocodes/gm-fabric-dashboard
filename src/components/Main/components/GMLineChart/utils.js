// importing a single function instead of the whole _
import isEmpty from "lodash/isEmpty";

/**
 * Converts timeseries with time stamps to a string of values without timestamps
 *
 * @param {array} array a timeseries array with time stamps
 * @returns {array}     an array of values
 */
export function numericalTimeSeriesFunc(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array[0].map(cell => cell[1]);
  } else {
    return 0;
  }
}

/**
 * Produces an average of array of values
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function average(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array.reduce((a, b) => a + b) / array.length;
  } else {
    return 0;
  }
}

/**
 * Produces a median of array of values
 * (if no middle value, average of 2 most central values is returned)
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function median(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    const len = array.length;
    let median = 0;
    array.sort();

    if (len % 2 === 0) {
      // if even, average of two middle numbers
      median = (array[len / 2 - 1] + array[len / 2]) / 2;
    } else {
      // if odd, middle number only
      median = array[(len - 1) / 2];
    }
    return median;
  } else {
    return 0;
  }
}

/**
 * Produces a mode(s) of array of values
 * if more than one mode, returns all applicable modes as well as their frequency
 *
 * @param {array} array an array of values
 * @returns {array}     an array that contains mode(s) and frequency value
 */
export function modes(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    let modes = [];
    let count = [];
    let i;
    let number;
    let maxIndex = 0;

    for (let i = 0; i < array.length; i++) {
      number = array[i];
      count[number] = (count[number] || 0) + 1;
      if (count[number] > maxIndex) {
        maxIndex = count[number];
      }
    }

    for (i in count)
      if (count.hasOwnProperty(i)) {
        if (count[i] === maxIndex) {
          modes.push(Number(i));
        }
      }

    modes.push({ frequency: maxIndex });

    if (maxIndex > 1) {
      return modes;
    } else {
      modes = "no value occurs more than once";
      return modes;
    }
  } else {
    return 0;
  }
}

/**
 * Produces a string that contains descriptive statistics of the graph
 *
 * @param {array} timeSeries an array of values
 * @param {string} title chart title
 * @returns {string}     returns a string with summary chart statistics
 */
export function screenReaderGraphDescription(timeSeries, title, intl) {
  const numericalTimeSeries = numericalTimeSeriesFunc(timeSeries);

  // We don't have negative values in our charts, thus average of 0 means that all values are 0, aka no meaningful data
  if (average(numericalTimeSeries) > 0) {
    let screenReaderGraphData = {
      median: median(numericalTimeSeries).toFixed(2),
      average: average(numericalTimeSeries).toFixed(2),
      mode: modes(numericalTimeSeries),
      max: numericalTimeSeries.sort()[numericalTimeSeries.length - 1],
      min: numericalTimeSeries.sort()[0],
      dataPoints: numericalTimeSeries.length,
      timeSeries: numericalTimeSeries,
      title
    };

    if (
      typeof screenReaderGraphData.mode !== "string" &&
      screenReaderGraphData.mode.length > 1
    ) {
      screenReaderGraphData = Object.assign({}, screenReaderGraphData, {
        mode: `Mode has frequency of ${
          screenReaderGraphData.mode.pop().frequency
        } and values of ${screenReaderGraphData.mode}`
      });
    }

    return intl.formatMessage(
      {
        id: "GMLineChart.screenReaderGraphDescription.withData",
        defaultMessage:
          "A tabular representation of the {title} chart data: median {median} average {average} mode {mode} maximum {max} minimum {min} number of observations {dataPoints} complete data time series follows {timeSeries}",
        description: "Screen reader description for GMLineChart"
      },
      screenReaderGraphData
    );
  } else {
    return intl.formatMessage({
      id: "GMLineChart.screenReaderGraphDescription.noData",
      defaultMessage: `The average for currently displayed data is equal to 0.`,
      description: "Screen reader description for GMLineChart"
    });
  }
}
