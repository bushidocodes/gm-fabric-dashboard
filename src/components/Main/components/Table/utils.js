import React from "react";
import TableColHeader from "components/Main/components/TableColHeader";
import TableColLatencyHeader from "components/Main/components/TableColLatencyHeader";

/** function getTableHeaders
 * takes Table type and returns appropriate headers with styling
 * for Function/Routes Table or GMServiceTable (Instances Table)
 * @param {object} items
 * @returns {React.element}
 */
export const getTableHeaders = type => {
  let headerNames = [];
  if (type === "Instance") headerNames = [type, "Uptime"];
  if (type === "Route" || type === "Function")
    headerNames = [type, "Requests/sec", "Requests", "Error %", "Latency"];
  let styleProps = {};

  // apply greater flex-basis for first column (20%)
  // requests/sec is a sparkline header and set to align left
  // set text-align to right for additional text columns
  return headerNames.map((header, headerIdx) => {
    styleProps = {};
    if (headerIdx === 0) styleProps = { flex: "1 1 20%" };
    if (headerIdx !== 0 && header !== "Requests/sec")
      styleProps = { textAlign: "right" };

    if (header === "Latency") {
      return <TableColLatencyHeader key={header} />;
    } else {
      return (
        <TableColHeader style={styleProps} key={header}>
          {header}
        </TableColHeader>
      );
    }
  });
};

// simple helper that evaluates and returns route or func props
export const getItem = (route, func) => {
  if (route !== undefined) return route;
  if (func !== undefined) return func;
};
