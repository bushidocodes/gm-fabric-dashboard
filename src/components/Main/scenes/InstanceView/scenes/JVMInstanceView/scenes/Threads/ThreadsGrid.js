import { Actions } from "jumpstate";
import { PropTypes } from "prop-types";
import React, { Component } from "react";
import { connect } from "react-redux";
import _ from "lodash";

import ThreadsTable from "./components/ThreadsTable";
import TableToolbar from "components/Main/components/TableToolbar";
import ErrorBoundary from "components/ErrorBoundary";
import { getVisibleThreads } from "utils/jvm/selectors";
import NotFoundError from "components/Main/components/NotFoundError";

/**
 * Parent container of ThreadsTable and TableToolbar
 * @class ThreadsGrid
 * @extends {Component}
 */
class ThreadsGrid extends Component {
  static propTypes = {
    fabricServer: PropTypes.string,
    selectedInstance: PropTypes.string,
    selectedService: PropTypes.string,
    selectedServiceVersion: PropTypes.string,
    threads: PropTypes.array,
    threadsEndpoint: PropTypes.string,
    threadsError: PropTypes.object
  };

  // Options for sort dropdown rendered in TableToolbar
  static sortByOptions = [
    {
      value: "state",
      label: "State"
    },
    {
      value: "name",
      label: "Name"
    },
    {
      value: "id",
      label: "ID"
    }
  ];

  static groupByOptions = [
    {
      value: "state",
      label: "State"
    },
    {
      value: "none",
      label: "None"
    }
  ];

  constructor(props) {
    super(props);
    this.state = {
      filterString: "",
      groupByAttribute: "none",
      sortByAttribute: "id",
      ascending: true
    };
  }

  componentDidMount() {
    const {
      fabricServer,
      selectedService,
      selectedServiceVersion,
      selectedInstance,
      threadsEndpoint
    } = this.props;

    // If fabricServer is truthy, we are running with a "Fabric Server" discovery service,
    // so we need to dynamically build the endpoint for the threads API. If falsy, we can just
    // pull the static configuration value from Redux via threadsEndpoint.
    // Technically Actions.fetchThreads defaults to the static threadsEndpoint, so this is
    // redundant, but we are showing it here just to be more explicit about the logic.

    if (this.props.fabricServer) {
      Actions.fetchAndStoreInstanceThreads(
        `${fabricServer}/threads/${selectedService}/${selectedServiceVersion}/${selectedInstance}`
      );
    } else {
      Actions.fetchAndStoreInstanceThreads(threadsEndpoint);
    }
  }

  /**
   * Helper function that takes the threads passed as props
   * and sorts according to how sortByAttribute and ascending
   * are set in the local state object.
   * @param {Array} threads
   */
  sort = threads => {
    const { sortByAttribute, ascending } = this.state;
    const sortOrder = ascending ? ["asc"] : ["desc"];
    // thread["id"] is a string, so we need to convert to an int to sort properly
    const sortFunc = thread => {
      return sortByAttribute === "id"
        ? parseInt(thread["id"], 10)
        : thread[sortByAttribute].toLowerCase();
    };
    return _.orderBy(threads, sortFunc, sortOrder);
  };

  setFilterString = filterString => this.setState({ filterString });

  setSortByAttribute = sortByAttribute => {
    if (sortByAttribute === this.state.sortByAttribute) {
      this.setState({ ascending: !this.state.ascending });
    } else {
      this.setState({ sortByAttribute });
    }
  };

  setGroupByAttribute = groupByAttribute => this.setState({ groupByAttribute });

  render() {
    const { threads, threadsError } = this.props;
    const filteredThreads = threads.filter(
      thread =>
        thread.name
          .toLowerCase()
          .indexOf(this.state.filterString.trim().toLowerCase()) !== -1
    );

    if (threads && threads.length > 0) {
      return (
        <ErrorBoundary>
          <TableToolbar
            searchInputProps={{
              filterString: this.state.filterString,
              setFilterString: this.setFilterString,
              searchPlaceholder: "Search Threads"
            }}
            sortByProps={{
              sortByOptions: ThreadsGrid.sortByOptions,
              sortByAttribute: this.state.sortByAttribute,
              setSortByAttribute: this.setSortByAttribute
            }}
            groupByProps={{
              groupByOptions: ThreadsGrid.groupByOptions,
              groupByAttribute: this.state.groupByAttribute,
              setGroupByAttribute: this.setGroupByAttribute
            }}
          />
          <ThreadsTable
            groupByAttribute={this.state.groupByAttribute}
            filteredThreadData={this.sort(filteredThreads)}
          />
        </ErrorBoundary>
      );
    } else if (_.isEmpty(threadsError)) {
      return <NotFoundError errorMsg="No Threads Found" />;
    } else {
      return <NotFoundError errorMsg="Failed to Fetch Threads" />;
    }
  }
}

function mapStateToProps(state) {
  return {
    fabricServer: state.settings.fabricServer,
    threads: getVisibleThreads(state),
    threadsError: state.instance.threadsError,
    threadsEndpoint: state.settings.threadsEndpoint,
    selectedService: state.fabric.selectedService,
    selectedServiceVersion: state.fabric.selectedServiceVersion,
    selectedInstance: state.fabric.selectedInstance
  };
}

export default connect(mapStateToProps)(ThreadsGrid);
