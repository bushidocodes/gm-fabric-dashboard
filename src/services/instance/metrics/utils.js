import { getState } from "jumpstate";
import { getFabricServer } from "utils/head";

export function buildDiscoveryServiceInstanceMetricsEndpoint(
  fabricServer = getState().fabric.fabricServer || getFabricServer(),
  service = getState().fabric.selectedService || "",
  version = getState().fabric.selectedServiceVersion || "",
  instanceID = getState().fabric.selectedInstance
) {
  return `${fabricServer}/metrics/${service}/${version}/${instanceID}`;
}

/**
 * Clears the interval with the ID stored at window.refreshInstanceMetricsIntervalID and then
 * wipes window.refreshMetricsIntervalID
 */
export function clearInstanceMetricsPollingIntervalIfNeeded() {
  if (window.refreshInstanceMetricsPollingInterval) {
    clearInterval(window.refreshInstanceMetricsPollingInterval);
    window.refreshInstanceMetricsPollingInterval = null;
  }
}
