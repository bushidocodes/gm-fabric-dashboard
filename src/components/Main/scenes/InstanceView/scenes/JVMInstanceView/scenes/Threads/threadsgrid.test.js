import React from "react";
import { Actions } from "jumpstate";
import configureMockStore from "redux-mock-store";
import { CreateJumpstateMiddleware } from "jumpstate";

// Utilities
import mockState from "json/mockReduxState";
import { mountWithIntl, renderWithIntl } from "utils/i18nTesting";

// Components
import ThreadsGrid from "./index";

// Mock the jumpstate effect used in ThreadsGrid
Actions.fetchAndStoreInstanceThreads = jest.fn();

// Create a mock store and initialize with mock data
const store = configureMockStore([CreateJumpstateMiddleware()])(mockState);

const threadsTableProps = {
  filteredThreadData: [
    {
      daemon: true,
      id: "2",
      name: "Test Runnable",
      priority: 8,
      stack: [
        "java.lang.Object.wait(Native Method)",
        "java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)",
        "java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)",
        "java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)"
      ],
      state: "RUNNABLE"
    }
  ]
};

// Wrap ThreadsGrid in Memory Router
const RouterWrap = <ThreadsGrid store={store} />;

let wrapper;

describe("ThreadsGrid View", () => {
  beforeEach(() => {
    wrapper = mountWithIntl(RouterWrap);
  });

  test("matches snapshot", () => {
    const tree = renderWithIntl(RouterWrap);
    expect(tree).toMatchSnapshot();
  });

  test("renders correct components", () => {
    expect(wrapper.find("TableToolbar")).toHaveLength(1);
    expect(wrapper.find("ThreadsTable")).toHaveLength(1);
    expect(wrapper.find("ErrorBoundary")).toHaveLength(1);
  });

  test("passes the correct props down to TableToolbar", () => {
    expect(wrapper.find("TableToolbar").props()).toMatchObject({
      groupByProps: {
        groupByAttribute: "none",
        groupByOptions: [
          { label: "State", value: "state" },
          { label: "None", value: "none" }
        ]
      },
      searchInputProps: {
        filterString: "",
        searchPlaceholder: "Search Threads"
      },
      sortByProps: {
        sortByAttribute: "id",
        sortByOptions: [
          { label: "State", value: "state" },
          { label: "Name", value: "name" },
          { label: "ID", value: "id" }
        ]
      }
    });
  });

  test("passes the correct props down to ThreadsTable", () => {
    expect(wrapper.find("ThreadsTable").props()).toMatchObject(
      threadsTableProps
    );
  });
});
