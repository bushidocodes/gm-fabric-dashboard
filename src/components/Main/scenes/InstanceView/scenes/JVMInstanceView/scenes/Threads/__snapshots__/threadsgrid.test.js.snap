// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadsGrid View matches snapshot 1`] = `
.c8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: rgb(255,255,255);
  box-shadow: inset 0 -1px hsl(0,0%,92%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-weight: 500;
  overflow: visible;
  width: 100%;
}

.c16 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
}

.c10 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c10:first-child {
  padding-left: 16px;
}

.c9 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-flex: 0 1 5em;
  -ms-flex: 0 1 5em;
  flex: 0 1 5em;
}

.c9:first-child {
  padding-left: 16px;
}

.c11 {
  display: block;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c13 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  -webkit-flex: 0 1 5em;
  -ms-flex: 0 1 5em;
  flex: 0 1 5em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
}

.c17 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  -webkit-flex: 0 1 5em;
  -ms-flex: 0 1 5em;
  flex: 0 1 5em;
  -webkit-flex: 0 1 7em;
  -ms-flex: 0 1 7em;
  flex: 0 1 7em;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  text-align: right;
}

.c12 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: rgb(255,255,255);
  box-shadow: inset 0 -1px hsl(0,0%,92%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  min-height: 36px;
  width: 100%;
  cursor: pointer;
}

.c12:hover {
  background-color: hsl(0,0%,98%);
}

.c12:focus {
  outline: -webkit-focus-ring-color auto 5px;
}

.c12:focus:active:hover {
  outline: 0;
}

.c12 > * {
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c12 > *:first-child {
  padding-left: 16px;
}

.c15 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c14 {
  z-index: 1;
  height: 24px;
  width: 24px;
  color: rgb(10,171,42);
  word-spacing: 0;
  vertical-align: middle;
}

.c14:not(:root) {
  overflow: visible;
}

.c18 {
  z-index: 1;
  height: 24px;
  width: 24px;
  color: currentColor;
  word-spacing: 0;
  vertical-align: middle;
}

.c18:not(:root) {
  overflow: visible;
}

.c7 {
  display: block;
  font-family: Rubik,-apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue",sans-serif;
  margin-left: auto;
  margin-right: auto;
  max-width: 1300px;
  width: 100%;
}

.c2 {
  background-image: url(magnifier.svg);
  background-repeat: no-repeat;
  background-size: contain;
  position: relative;
  text-indent: 20px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: black;
  font-family: Nunito,-apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue",sans-serif;
  font-weight: 600;
  font-size: 14px;
  line-height: 1.4;
  text-align: left;
  text-transform: none;
  padding: 1.4px 4px;
  background-color: rgb(255,255,255);
  border-radius: 3px;
  border: 1px solid rgb(242,242,242);
  border-bottom-color: rgb(230,230,230);
  -webkit-user-select: auto;
  -moz-user-select: auto;
  -ms-user-select: auto;
  user-select: auto;
  box-shadow: inset 0 0 0 transparent,0 0 0 transparent,0 0 0 transparent;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

.c2::-webkit-search-cancel-button {
  cursor: pointer;
}

.c2:hover {
  background-color: rgb(250,250,250);
  border-color: rgb(237,237,237);
  -webkit-transition: all 0.1s ease;
  transition: all 0.1s ease;
  border-bottom-color: rgb(217,217,217);
  color: black;
}

.c2:focus {
  outline: none;
  box-shadow: 0 0 0 2px rgb(10,171,42);
  z-index: 1;
}

.c2::-webkit-input-placeholder {
  color: currentColor;
  opacity: 1;
  text-align: center;
}

.c2::-moz-placeholder {
  color: currentColor;
  opacity: 1;
  text-align: center;
}

.c2:-ms-input-placeholder {
  color: currentColor;
  opacity: 1;
  text-align: center;
}

.c2::placeholder {
  color: currentColor;
  opacity: 1;
  text-align: center;
}

.c2::-ms-expand {
  background-color: transparent;
  border: 0;
}

.c5 {
  font-family: "Nunito",-apple-system,BlinkMacSystemFont,"Segoe UI","Roboto", "Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-flex-grow: 1;
  -ms-flex-grow: 1;
  flex-grow: 1;
  font-size: 14px;
  font-weight: 600;
  height: 28px;
  max-width: 125px;
  position: relative;
  width: 100%;
  z-index: 1020;
}

.c5 .Select-control {
  height: 28px !important;
}

.c5 .Select-control .Select-placeholder {
  line-height: 28px;
}

.c5 .Select-control .Select-value {
  line-height: 28px !important;
}

.c5 .Select-control .Select-value .Select-value-label {
  line-height: 28px;
}

.c5 .Select-control .Select-input {
  height: 28px;
}

.c6 {
  margin-left: 4px;
  opacity: 0.6;
  font-size: 11px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: hsl(0,0%,96%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 36px;
  margin-left: auto;
  margin-right: auto;
  max-width: 1300px;
  padding: 2px 16px;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media all and (min-width:1300px) {
  .c0:after {
    background-color: inherit;
    content: "";
    height: inherit;
    left: 0;
    pointer-events: none;
    position: absolute;
    right: 0;
    top: 0;
    z-index: 1;
  }

  .c0 > * {
    position: relative;
    z-index: 2;
  }
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          aria-label="Search Threads"
          className="c2"
          onChange={[Function]}
          placeholder="Search Threads"
          type="search"
          value=""
        />
      </form>
    </div>
    <div
      className="c3"
    />
    <div
      className="c4"
    >
      <div
        className="Select c5 Select--single is-searchable has-value"
        style={undefined}
      >
        <input
          disabled={false}
          name="form-field-group-by"
          type="hidden"
          value="none"
        />
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={undefined}
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-4--value"
          >
            <div
              className="Select-value"
              style={undefined}
              title={undefined}
            >
              <span
                aria-selected="true"
                className="Select-value-label"
                id="react-select-4--value-item"
                role="option"
              >
                <span>
                  <span>
                    Group
                     
                  </span>
                  <span
                    className="c6"
                  >
                    None
                  </span>
                </span>
              </span>
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-4--value"
                aria-describedby={undefined}
                aria-expanded="false"
                aria-haspopup="false"
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className={undefined}
                id={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                tabIndex={undefined}
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
      <div
        className="Select c5 Select--single is-searchable has-value"
        style={undefined}
      >
        <input
          disabled={false}
          name="form-field-sort-by"
          type="hidden"
          value="id"
        />
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={undefined}
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-5--value"
          >
            <div
              className="Select-value"
              style={undefined}
              title={undefined}
            >
              <span
                aria-selected="true"
                className="Select-value-label"
                id="react-select-5--value-item"
                role="option"
              >
                <span>
                  <span>
                    Sort
                     
                  </span>
                  <span
                    className="c6"
                  >
                    ID
                  </span>
                </span>
              </span>
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-5--value"
                aria-describedby={undefined}
                aria-expanded="false"
                aria-haspopup="false"
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className={undefined}
                id={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                tabIndex={undefined}
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c7"
  >
    <div
      className="c8"
    >
      <div
        className="c9"
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        State
      </div>
      <div
        className="c9"
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        ID
      </div>
      <div
        className="c10"
      >
        Name
      </div>
      <div
        className="c9"
        style={
          Object {
            "flex": "0 1 7em",
            "textAlign": "right",
          }
        }
      >
        Daemon
      </div>
      <div
        className="c9"
        style={
          Object {
            "flex": "0 1 7em",
            "textAlign": "right",
          }
        }
      >
        Priority
      </div>
      <div
        className="c9"
      />
    </div>
    <ol
      className="c11"
    >
      <li
        className="c12"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="link"
        tabIndex={0}
      >
        <div
          className="c13"
        >
          <svg
            aria-labelledby="ariaLabelledby"
            className="c14"
            focusable="false"
            version="1.1"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
          >
            <g
              className="iconBackground"
              fill="rgb(10, 171, 42)"
              fillOpacity=".3"
              fillRule="evenodd"
              id="Square"
            >
              <path
                d="M12,21 C7.02943725,21 3,16.9705627 3,12 C3,7.02943725 7.02943725,3 12,3 C16.9705627,3 21,7.02943725 21,12 C21,16.9705627 16.9705627,21 12,21 Z"
                id="BackgroundCircleSmall"
              />
            </g>
            <g
              className="c15"
              fill="currentColor"
              title=""
            >
              <title>
                
              </title>
              <path
                d="M11,9.86851709 L11,14.1314829 L14.1972244,12 L11,9.86851709 Z M16,12 L10,16 L10,8 L16,12 Z"
                id="GlyphRunningSmall"
              />
            </g>
          </svg>
        </div>
        <div
          className="c13"
        >
          2
        </div>
        <div
          className="c16"
        >
          Test Runnable
        </div>
        <div
          className="c17"
        >
          Yes
        </div>
        <div
          className="c17"
        >
          8
        </div>
        <div
          className="c13"
        >
          <svg
            aria-labelledby="ariaLabelledby"
            className="c18"
            focusable="false"
            version="1.1"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
          >
            <g
              className="c15"
              fill="currentColor"
              title=""
            >
              <title>
                
              </title>
              <g
                className="glyph"
                fill="currentColor"
                transform="translate(0 0) scale(1)"
              >
                <path
                  d="M7,7 L19,7 L19,9 L7,9 L7,7 Z M6,11 L18,11 L18,13 L6,13 L6,11 Z M5,15 L17,15 L17,17 L5,17 L5,15 Z"
                  id="shape"
                />
              </g>
            </g>
          </svg>
        </div>
      </li>
    </ol>
  </div>
</div>
`;
